libraries/map/src/main/kotlin/io/element/android/libraries/map/LocationUpdatesFlow.kt

```diff
@@ -0,0 +1,101 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.map
+
+import android.Manifest
+import android.content.Context
+import android.location.LocationManager
+import androidx.annotation.RequiresPermission
+import androidx.core.content.getSystemService
+import androidx.core.location.LocationListenerCompat
+import androidx.core.location.LocationManagerCompat
+import androidx.core.location.LocationRequestCompat
+import io.element.android.libraries.core.coroutine.CoroutineDispatchers
+import kotlinx.coroutines.asExecutor
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.trySendBlocking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.callbackFlow
+
+/**
+ * Returns a cold [Flow] that, once collected, emits [Location] updates every second.
+ */
+@RequiresPermission(
+    anyOf = [
+        Manifest.permission.ACCESS_COARSE_LOCATION,
+        Manifest.permission.ACCESS_FINE_LOCATION
+    ]
+)
+fun locationUpdatesFlow(
+    context: Context,
+    coroutineDispatchers: CoroutineDispatchers,
+): Flow<Location> = callbackFlow {
+    val locationManager: LocationManager = checkNotNull(context.getSystemService())
+    val provider = locationManager.bestAvailableProvider()
+    // Try to eagerly emit the last known location as fast as possible
+    locationManager.getLastKnownLocation(provider)?.let { location ->
+        trySendBlocking(
+            Location(
+                lat = location.latitude,
+                lon = location.longitude,
+                accuracy = location.accuracy
+            )
+        )
+    }
+    val locationListener = LocationListenerCompat { location ->
+        trySendBlocking(
+            Location(
+                lat = location.latitude,
+                lon = location.longitude,
+                accuracy = location.accuracy
+            )
+        )
+    }
+    LocationManagerCompat.requestLocationUpdates(
+        locationManager,
+        provider,
+        buildLocationRequest(),
+        coroutineDispatchers.io.asExecutor(),
+        locationListener,
+    )
+    awaitClose {
+        LocationManagerCompat.removeUpdates(locationManager, locationListener)
+    }
+}
+
+data class Location(
+    val lat: Double,
+    val lon: Double,
+    val accuracy: Float,
+)
+
+private fun LocationManager.bestAvailableProvider(): String =
+    checkNotNull(getProviders(true).minByOrNull { providerPriority(it) }) {
+        "No location provider available"
+    }
+
+private fun providerPriority(provider: String): Int = when (provider) {
+    LocationManager.FUSED_PROVIDER -> 0
+    LocationManager.GPS_PROVIDER -> 1
+    LocationManager.NETWORK_PROVIDER -> 2
+    LocationManager.PASSIVE_PROVIDER -> 3
+    else -> 4
+}
```

Comment:
nit: this would make a lot more sense to me if the priority order was reversed (higher = better)

Reply:
I was thinking about linux's process priorities (in which 0 is highest priority) but no probs if you wanna reverse it.

Reply:
I'd definitely prefer the reverse :grin: 

Linux priority is weird because (as I understand it) you're actually setting the "niceness" of the process, which is effectively the inverse of priority (if you're very nice you cede all the resources to other processes).

Reply:
Allrighty! Here you go!